pig/src/main/java/org/jboss/pnc/bacon/pig/impl/addons/scanservice/PostBuildScanService.java

```diff
@@ -0,0 +1,222 @@
+import org.jboss.pnc.bacon.pig.impl.addons.scanservice.pssaas.ScanHelper;
+import org.jboss.pnc.bacon.pig.impl.config.PigConfiguration;
+import org.jboss.pnc.bacon.pig.impl.pnc.PncBuild;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static java.lang.String.*;
+
+public class PostBuildScanService extends AddOn implements Validate {
+    private static final Logger log = LoggerFactory.getLogger(PostBuildScanService.class);
+    private static final String ADDON_SCAN_SERVICE_MAP_KEY = "postBuildScanService";
+    public static final String KEY_VAL_PAIR_FMT = "%18s : %s";
+
+    private ScanServiceDTO scanServiceDTO;
+
+    private static final Set<String> mandatoryItems = new HashSet<>();
+    static {
+        mandatoryItems.add("productID");
+        mandatoryItems.add("serviceUrl");
+        mandatoryItems.add("serviceSecretKey");
+        mandatoryItems.add("serviceSecretValue");
+    }
+
+    private static final Set<String> masterSet = new HashSet<>();
+    static {
+        masterSet.add("productID");
+        masterSet.add("eventId");
+        masterSet.add("isManagedService");
+        masterSet.add("cpaasVersion");
+        masterSet.add("jobUrl");
+        masterSet.add("serviceUrl");
+        masterSet.add("serviceSecretKey");
+        masterSet.add("serviceSecretValue");
+        masterSet.add("brewBuilds");
+        masterSet.add("extraScmUrls");
+        masterSet.add("extraBuilds");
+        masterSet.add("scmUrl");
+        masterSet.add("scmRevision");
+        masterSet.add("name");
+    }
+
+    Map<String, ?> postBuildScanConfigMap;
+
+    public PostBuildScanService(
+            PigConfiguration pigConfiguration,
+            Map<String, PncBuild> builds,
+            String releasePath,
+            String extrasPath) {
+        super(pigConfiguration, builds, releasePath, extrasPath);
+    }
+
+    @Override
+    protected String getName() {
+        return ADDON_SCAN_SERVICE_MAP_KEY;
+    }
+
+    @Override
+    public void trigger() {
+        log.info("=============== TRIGGER entry point for REST call ===============");
+        log.info("Addon Configuration Parameters");
+        Map<String, Map<String, ?>> addonsMap = pigConfiguration.getAddons();
+        postBuildScanConfigMap = addonsMap.get(ADDON_SCAN_SERVICE_MAP_KEY);
+        this.validate();
+        scanServiceDTO = createScanServiceDTO();
+        sendScanRequest(scanServiceDTO);
+
+        ScanHelper pssaasService = null;
+        Response response = null;
+        try {
+            pssaasService = new ScanHelper(
+                    this.getServiceSecretKey(),
+                    this.getServiceSecretValue(),
+                    this.getServiceUrl());
+            pssaasService.client.triggerScan(scanServiceDTO);
```

Comment:
The original code here that contained `response.bufferEntity();` was to work around an issue we found with the underlying Apache HTTP Client not following redirects when `setFollowRedirects(true)` for `POST` requests, we could see that it was being set but seemingly `ApacheHttpClient43Engine` only follows redirects for `GET` request, the [spec](https://www.rfc-editor.org/rfc/rfc7231#section-6.4.3) is a bit ambiguous as to if this is the correct behaviour.

We'll likely never see a 30x response with the intended service but other services may do so.
We've also added the logging to the client itself, we don't actually do anything with the response from the service, but if we do need to then we will create a response Pojo.

The service will currently return a UUID but there is nothing further that can be done with this UUID currently (no API endpoints accept this for "status" or scan result for example), we simply log the response body using the filters in `RedirectAndLog`, even if the scan is invalid or fails immediately then the same response is received.

I suspect this will change in future but we can create the response Pojo at this point, until then its basically a fire and forget POST request.

Example return - 
`{"eventListener":"pssc-scan-listener","namespace":"pssaas-service", "eventListenerUID":"s0me-UUID","eventID":"an0Ther-UID"}`

Reply:
Thanks for the explanation. Perhaps we should give _some_ feedback to the user with e.g. `log.info("PSSaS Scan sent successfully (UUID: .... )"` or whether there was an error code in transmission.

I think the try-with-resources still applies.

Reply:
```suggestion
            Response response = pssaasService.client.triggerScan( scanServiceDTO);
            response.close();
```